bundle/tpm2_bundle.c

```diff
@@ -0,0 +1,219 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <openssl/err.h>
+#include <openssl/evp.h>
+
+#include <tss2/tss2_tctildr.h>
+
+#include "log.h"
+#include "tpm2_errata.h"
+#include "tpm2_options.h"
+#include "tpm2_tool.h"
+#include "tpm2_tool_output.h"
+
+#include "gen-bundle.h"
+
+typedef struct {
+	const char * name;
+	bool (*onstart)(tpm2_options **opts);
+	tool_rc (*onrun)(ESYS_CONTEXT *ectx, tpm2_option_flags flags);
+	tool_rc (*onstop)(ESYS_CONTEXT *ectx);
+	void (*onexit)(void);
+} tpm2_tool_t;
+
+static const tpm2_tool_t tpm2_tools[] = {
+#include "gen-bundle.c"
+};
+
+#define SUPPORTED_ABI_VERSION \
+{ \
+    .tssCreator = 1, \
+    .tssFamily = 2, \
+    .tssLevel = 1, \
+    .tssVersion = 108, \
+}
+
+static void esys_teardown(ESYS_CONTEXT **esys_context) {
+
+    if (esys_context == NULL)
+        return;
+    if (*esys_context == NULL)
+        return;
+    Esys_Finalize(esys_context);
+}
+
+static void teardown_full(ESYS_CONTEXT **esys_context) {
+
+    TSS2_TCTI_CONTEXT *tcti_context = NULL;
+    TSS2_RC rc;
+
+    if (!*esys_context) {
+        return;
+    }
+
+    rc = Esys_GetTcti(*esys_context, &tcti_context);
+    if (rc != TPM2_RC_SUCCESS)
+        return;
+    esys_teardown(esys_context);
+    Tss2_TctiLdr_Finalize(&tcti_context);
+}
+
+static ESYS_CONTEXT* ctx_init(TSS2_TCTI_CONTEXT *tcti_ctx) {
+
+    TSS2_ABI_VERSION abi_version = SUPPORTED_ABI_VERSION;
+    ESYS_CONTEXT *esys_ctx;
+
+    TSS2_RC rval = Esys_Initialize(&esys_ctx, tcti_ctx, &abi_version);
```

Comment:
You can just pass NULL for the ABI version. I'm assuming this is a copy paste, so we're doing it in tools.

Reply:
Right, that's a direct copy of the `tools/tpm2_tool.c` program: https://github.com/tpm2-software/tpm2-tools/blob/master/tools/tpm2_tool.c#L52